lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -1906,6 +1906,30 @@ void GrandCentralPass::runOnOperation() {
   auto instancePathCache = InstancePathCache(getAnalysis<InstanceGraph>());
   instancePaths = &instancePathCache;
 
+  /// Contains the set of modules which are instantiated by the DUT, but not a
+  /// companion or instantiated by a companion.  If no DUT exists, treat the top
+  /// module as if it were the DUT.  This works by doing a depth-first walk of
+  /// the instance graph, starting from the "effective" DUT and stopping the
+  /// search at any modules which are known companions.
+  DenseSet<hw::HWModuleLike> dutModules;
+  FModuleOp effectiveDUT = dut;
+  if (!effectiveDUT)
+    effectiveDUT =
+        cast<FModuleOp>(instancePaths->instanceGraph.getTopLevelModule());
+  auto dfRange =
+      llvm::depth_first(instancePaths->instanceGraph.lookup(effectiveDUT));
+  for (auto i = dfRange.begin(), e = dfRange.end(); i != e;) {
+    auto module = cast<FModuleLike>(*i->getModule());
+    if (AnnotationSet(module).hasAnnotation(companionAnnoClass)) {
+      i.skipChildren();
+      continue;
+    }
+    dutModules.insert(i->getModule());
+    // Manually increment the iterator to avoid walking off the end from
+    // skipChildren.
+    ++i;
+  }
```

Comment:
I'm using the explicit `dfRange` as opposed to `llvm::depth_first` because I need to get at the underlying `df_iterator` in order to access the `skipChildren` method. I tried to do this with `llvm::depth_first`, but couldn't figure out how to do it. 😕 Maybe you have some idea on a cleaner approach here @youngar.

Reply:
Looks good to me, I don't think there is any other way to get at that `skipChildren` function.